### PR TITLE
Enhancements for AGG IF to FILTER rewrite

### PR DESCRIPTION
**Pull request by zacw7**
This PR has the following 3 commits in order:
1. `Add a new rewrite strategy UNWRAP_IF_SAFE`
   - A safe version of `UNWRAP_IF`. In this mode, we only unwrap IF for cases which we are sure its 100% safe to do it.
   - Update the corresponding testing to adopt this change.
2. `Add support for AGG(CAST(IF(expr, x))`
   - Improve the rule to handle this case when CAST() is present by rewriting AGG(CAST(IF(expr, x)) to AGG(CAST(x)) FILTER (WHERE expr).
   - Add a new unit test for CAST().
3. `Add a new stats for AGG(IF()) rewrite`
   - The new stats would be present in `RuntimeStats` if any node has been changed by the rewrite rule.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java [hunk 1]
```diff
@@ -231,17 +250,22 @@ private boolean shouldRewriteAggregation(Aggregation aggregation, ProjectNode so
...
         Set<VariableReferenceExpression> ifResultReferences = VariablesExtractor.extractUnique(ifExpression.getArguments().get(1));
         if (ifConditionReferences.stream().noneMatch(ifResultReferences::contains)) {
+            // The IF expression can be safely unwrapped, ignore the rewrite strategy.
             return true;
         }
 
+        if (rewriteStrategy == AggregationIfToFilterRewriteStrategy.FILTER_WITH_IF) {
+            return false;
+        }
+
```
Comment (yuanzhanhku) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java:268]: Could you add test where the two modes behave differently? such as:
IF(a > 1, a)

**Review by yuanzhanhku — COMMENTED** [review]
Nice change! Thanks.

**Comment by kaikalur** [comment]
I will look at the code later but at a high level, item 1, "apply in safe situations" is not advisable. If the config setting says don't do it, just don't do it.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java [hunk 2]
```diff
@@ -117,6 +118,7 @@ public Result apply(AggregationNode aggregationNode, Captures captures, Context
             return Result.empty();
         }
 
+        context.getSession().getRuntimeStats().addMetricValue(REWRITE_AGGREGATION_IF_TO_FILTER, 1);
         // Get the corresponding assignments in the input project.
         // The aggregationReferences only has the aggregations to rewrite, thus the sourceAssignments only has IF expressions with NULL false results.
```
[the remark this answers is not in the document]
Comment (zacw7) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java]: Good catch. Updated the class comment to reflect the new unwrapping logic also.

**Comment by zacw7** [comment]
> I will look at the code later but at a high level, item 1, "apply in safe situations" is not advisable. If the config setting says don't do it, just don't do it.

Thanks for the suggestion. Ideally we want to wrap `IF` as much as we, so IMO if we are sure 100% it's safe then we should do so. That said, I agree with you that the current strategy name might be a bit confusing, since in some cases it indicates filter _with if_ but we are actually unwrapping `IF`. Maybe we can reword the strategies into something else? Any thoughts? @yuanzhanhku

**Comment by yuanzhanhku** [comment]
> > I will look at the code later but at a high level, item 1, "apply in safe situations" is not advisable. If the config setting says don't do it, just don't do it.
> 
> Thanks for the suggestion. Ideally we want to unwrap `IF` as much as we, so IMO if we are sure it's 100% safe then we should do so. That said, I agree with you that the current strategy name might be a bit confusing, since in some cases it indicates filter _with if_ but we are actually unwrapping `IF`. Maybe we can reword the strategies into something else? Any thoughts? @yuanzhanhku

Unwrapping IF could give significant improvements compared with keeping the IF. We would like to have two modes for this optimization:
 - a safe version which could be enabled by default. In this mode, we only unwrap IF for cases which we are sure its 100% safe to do it.
 - an unsafe version which is only enabled for certain workloads like system generated queries with fixed query pattern. In this mode, we only keep IF for cases which we are sure it will cause issues if we unwrap it.

The current FILTER_WITH_IF name is misleading if we use it for the safe version of unwrapping IF. Maybe we could add another mode `UNWRAP_IF_SAFE` and leave FILTER_WITH_IF with its old behavior to always keep the IF, (maybe deprecate it later after we enabled `UNWRAP_IF_SAFE` everywhere without any issues).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java [hunk 3]
```diff
@@ -74,7 +77,10 @@
  * <p>
  * Note that unwrapping the IF expression in the aggregate might cause issues if the true branch return errors for rows not matching the filters. For example:
  * 'IF(CARDINALITY(array) > 0, array[1]))'
- * Session property AGGREGATION_IF_TO_FILTER_REWRITE_STRATEGY and canUnwrapIf() control whether to enable IF unwrapping.
+ * Session property AGGREGATION_IF_TO_FILTER_REWRITE_STRATEGY and canUnwrapIf() control whether to enable IF unwrapping:
+ * 1. If the strategy is FILTER_WITH_IF, then keep the IF expression.
+ * 2. If it is certain that the IF expression can be unwrapped safe, then do so.
```
Comment (yuanzhanhku) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java:82]: Mention that this is for UNWRAP_IF_SAFE mode.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java [hunk 4]
```diff
@@ -74,7 +77,10 @@
...
+ * 3. If the strategy is UNWRAP_IF, then unwrap the IF expression after it passes the remaining checks.
```
Comment (yuanzhanhku) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java:83]: Mention that this is an unsafe mode and the checks are not exhaustive. There may be cases where unwrapping the if could lead to errors.

**Review by yuanzhanhku — APPROVED** [review]
Looks great. Thanks!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java [hunk 5]
```diff
@@ -74,7 +77,10 @@
  * <p>
  * Note that unwrapping the IF expression in the aggregate might cause issues if the true branch return errors for rows not matching the filters. For example:
  * 'IF(CARDINALITY(array) > 0, array[1]))'
- * Session property AGGREGATION_IF_TO_FILTER_REWRITE_STRATEGY and canUnwrapIf() control whether to enable IF unwrapping.
+ * Session property AGGREGATION_IF_TO_FILTER_REWRITE_STRATEGY and canUnwrapIf() control whether to enable IF unwrapping:
+ * 1. If the strategy is FILTER_WITH_IF, then keep the IF expression.
+ * 2. If the strategy is UNWRAP_IF_SAFE/UNWRAP_IF and the IF expression can be unwrapped safely, then do so.
```
Comment (yuanzhanhku) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java:82]: UNWRAP_IF doesn't guarantee that the IF can be unwrapped safely. We can remove it from this bullet point.

**Review by yuanzhanhku — APPROVED** [review]

**Review by yuanzhanhku — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java [hunk 6]
```diff
@@ -228,6 +239,12 @@ private boolean shouldRewriteAggregation(Aggregation aggregation, ProjectNode so
             return false;
         }
         RowExpression sourceExpression = sourceProject.getAssignments().get((VariableReferenceExpression) aggregation.getArguments().get(0));
+        if (sourceExpression instanceof CallExpression
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java:244]: nit: break into two ifs
```java
if (sourceExpression instanceof CallExpression) {
    CallExpression callExpression = (CallExpression) sourceExpression;
    if (callExpression.getArguments().size() == 1

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java [hunk 7]
```diff
@@ -228,6 +239,12 @@ private boolean shouldRewriteAggregation(Aggregation aggregation, ProjectNode so
...
+                && ((CallExpression) sourceExpression).getArguments().size() == 1
+                && "CAST".equals(((CallExpression) sourceExpression).getDisplayName())) {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java:246]: Store `FunctionResolution` as a final member variable and use `functionResolution.isCastFunction(callExpression.getFunctionHandle())`

**Review by highker — APPROVED** [review]

**Pull request merged**